jx boot verify-jenkins-x-environment step fail on github enterprise

### Summary 

log
```
STEP: verify-jenkins-x-environment command: /bin/sh -c jx step verify env in dir: /Users/user/workspace/kubernetes/jenkins-x-boot-config

Storing the requirements in team settings in the dev environment
Validating git repository for dev environment at URL https://[my-github-enterprise-url]/[my-github-enterprise-user]/environment-[my-clusterName]-dev.git
error: handle dev environment repository: creating remote for dev environment environment-[my-clusterName]-dev: duplicating https://github.com/jenkins-x/jenkins-x-boot-config to [my-github-enterprise-user]/environment-[my-clusterName]-dev: getting repo for https://github.com/jenkins-x/jenkins-x-boot-config: failed to get repository jenkins-x/jenkins-x-boot-config due to: GET https://[my-github-enterprise]/api/v3/repos/jenkins-x/jenkins-x-boot-config: 404 Not Found []
error: failed to interpret pipeline file jenkins-x.yml: failed to run '/bin/sh -c jx step verify env' command in directory '.', output: ''
```

jx-requirements.yml
```
autoUpdate:
  enabled: false
  schedule: ""
bootConfigURL: https://github.com/jenkins-x/jenkins-x-boot-config.git
cluster:
  clusterName: [my-clusterName]
  environmentGitOwner: [my-github-enterprise-user]
  environmentGitPublic: true
  gitKind: github
  gitName: ghe
  gitPublic: true
  gitServer: https://[my-github-enterprise-url]
  namespace: jx
  provider: kubernetes
  registry: https://[my-harbor-url]/
environments:
- ingress:
    domain: [my-domain]
    externalDNS: true
    namespaceSubDomain: -jx.
    tls:
      email: ""
      enabled: false
      production: false
  key: dev
- ingress:
    domain: ""
    externalDNS: false
    namespaceSubDomain: ""
    tls:
      email: ""
      enabled: false
      production: false
  key: staging
- ingress:
    domain: ""
    externalDNS: false
    namespaceSubDomain: ""
    tls:
      email: ""
      enabled: false
      production: false
  key: production
gitops: true
ingress:
  domain: [my-domain]
  externalDNS: true
  namespaceSubDomain: -jx.
  tls:
    email: ""
    enabled: false
    production: false
kaniko: true
repository: nexus
secretStorage: local
storage:
  backup:
    enabled: false
    url: ""
  logs:
    enabled: false
    url: ""
  reports:
    enabled: false
    url: ""
  repository:
    enabled: false
    url: ""
vault: {}
velero: {}
versionStream:
  ref: v1.0.248
  url: https://github.com/jenkins-x/jenkins-x-versions.git
webhook: lighthouse

```

### Steps to reproduce the behavior

verify-jenkins-x-environment

### Expected behavior

success to get repository jenkins-x/jenkins-x-boot-config from https://github.com (not from [my-github-enterprise])

### Actual behavior

jx boot tries to find a repo in [my-github-enterprise]

### Jx version

The output of `jx version` is:

```
NAME               VERSION
jx                 2.0.1065
Kubernetes cluster v1.13.5
kubectl            v1.16.3
helm client        Client: v2.14.3+g0e7f3b6
git                2.20.1 (Apple Git-117)
Operating System   Mac OS X 10.14.6 build 18G87
```

### Jenkins type

<!--
Select which installation type are you using. 
-->
- [x] Serverless Jenkins X Pipelines (Tekton + Prow)
- [ ] Classic Jenkins

### Kubernetes cluster

<!--
What kind of Kubernetes cluster are you using & how did you create it?
-->

custom kubernetes (created and provided by our company for private)

### Operating system / Environment

<!--
In which environment are you running the jx CLI?
-->
Mac OS X 10.14.6 build 18G87

--- a/boot-vault.platform.yaml.template
+++ b/boot-vault.platform.yaml.template
@@ -21,7 +21,28 @@ jenkins:
         EnvVars:
           CODECOV_TOKEN: $CODECOV_TOKEN
 
+{{- if eq .Requirements.webhook "lighthouse" }}
 controllerbuild:
+  enabled: true
+  args:
+  - "controller"
+  - "build"
+  - "--git-reporting"
+  - "--batch-mode"
+  - "--git-credentials"
+  - "--verbose"
   image:
     repository: gcr.io/jenkinsxio/builder-go
     tag: $VERSION
+{{- else }}
+controllerbuild:
+  enabled: true
+  args:
+  - "controller"
+  - "build"
+  - "--batch-mode"
+  - "--verbose"
+  image:
+    repository: gcr.io/jenkinsxio/builder-go
+    tag: $VERSION
+{{- end }}

--- a/jenkins-x-boot-lh-ghe.yml
+++ b/jenkins-x-boot-lh-ghe.yml
@@ -1,0 +1,159 @@
+buildPack: none
+pipelineConfig:
+  pipelines:
+    pullRequest:
+      pipeline:
+        agent:
+          image: gcr.io/kaniko-project/executor:9912ccbf8d22bbafbf971124600fbb0b13b9cbd6
+        options:
+          containerOptions:
+            resources:
+              limits:
+                cpu: 4
+                memory: 6144Mi
+              requests:
+                cpu: 1
+                memory: 2048Mi
+        stages:
+          - name: build
+            environment:
+              - name: GIT_COMMITTER_EMAIL
+                value: jenkins-x@googlegroups.com
+              - name: GIT_AUTHOR_EMAIL
+                value: jenkins-x@googlegroups.com
+              - name: GIT_AUTHOR_NAME
+                value: jenkins-x-bot
+              - name: GIT_COMMITTER_NAME
+                value: jenkins-x-bot
+              - name: BASE_WORKSPACE
+                value: /workspace/source
+              - name: GOPROXY
+                value: http://jenkins-x-athens-proxy:80
+              - name: PARALLEL_BUILDS
+                value: "2"
+              - name: GOOGLE_APPLICATION_CREDENTIALS
+                value: /secrets/kaniko/kaniko-secret.json
+              - name: DISABLE_TEST_CACHING
+                value: "true"
+            options:
+              volumes:
+                - name: kaniko-secret
+                  secret:
+                    secretName: kaniko-secret
+                    items:
+                      - key: kaniko-secret
+                        path: kaniko/kaniko-secret.json
+              containerOptions:
+                volumeMounts:
+                  - name: kaniko-secret
+                    mountPath: /secrets
+
+            steps:
+              - name: build-binary
+                image: docker.io/golang:1.12.4
+                command: make
+                args: ['linux']
+
+              - name: validate-binary
+                image: docker.io/golang:1.12.4
+                command: "./build/linux/jx"
+                args: ['help']
+
+              - name: build-and-push-image
+                command: /kaniko/executor
+                args: ['--dockerfile=/workspace/source/Dockerfile','--destination=gcr.io/jenkinsxio/jx:${inputs.params.version}','--context=/workspace/source','--cache-repo=jenkins-x-docker-registry.jx.svc.cluster.local:5000/','--cache=true','--cache-dir=/workspace','--skip-tls-verify-registry=jenkins-x-docker-registry.jx.svc.cluster.local:5000']
+
+              - name: build-and-push-nodejs
+                command: /kaniko/executor
+                args: ['--dockerfile=/workspace/source/Dockerfile.builder-nodejs','--destination=gcr.io/jenkinsxio/builder-nodejs:${inputs.params.version}','--context=/workspace/source','--cache-repo=jenkins-x-docker-registry.jx.svc.cluster.local:5000/','--cache=true','--cache-dir=/workspace','--skip-tls-verify-registry=jenkins-x-docker-registry.jx.svc.cluster.local:5000']
+
+              - name: build-and-push-maven
+                command: /kaniko/executor
+                args: ['--dockerfile=/workspace/source/Dockerfile.builder-maven','--destination=gcr.io/jenkinsxio/builder-maven:${inputs.params.version}','--context=/workspace/source','--cache-repo=jenkins-x-docker-registry.jx.svc.cluster.local:5000/','--cache=true','--cache-dir=/workspace','--skip-tls-verify-registry=jenkins-x-docker-registry.jx.svc.cluster.local:5000']
+
+              - name: build-and-push-go
+                command: /kaniko/executor
+                args: ['--dockerfile=/workspace/source/Dockerfile.builder-go','--destination=gcr.io/jenkinsxio/builder-go:${inputs.params.version}','--context=/workspace/source','--cache-repo=jenkins-x-docker-registry.jx.svc.cluster.local:5000/','--cache=true','--cache-dir=/workspace','--skip-tls-verify-registry=jenkins-x-docker-registry.jx.svc.cluster.local:5000']
+
+              - name: build-and-push-go-nodejs
+                command: /kaniko/executor
+                args: ['--dockerfile=/workspace/source/Dockerfile.builder-go-nodejs','--destination=gcr.io/jenkinsxio/builder-go-nodejs:${inputs.params.version}','--context=/workspace/source','--cache-repo=jenkins-x-docker-registry.jx.svc.cluster.local:5000/','--cache=true','--cache-dir=/workspace','--skip-tls-verify-registry=jenkins-x-docker-registry.jx.svc.cluster.local:5000']
+
+          - name: e2e-tests
+            options:
+              volumes:
+                - name: sa
+                  secret:
+                    secretName: bdd-secret
+                    items:
+                      - key: bdd-credentials.json
+                        path: bdd/sa.json
+              containerOptions:
+                volumeMounts:
+                  - mountPath: /secrets
+                    name: sa
+            environment:
+              - name: GKE_SA
+                value: /secrets/bdd/sa.json
+              - name: GIT_COMMITTER_EMAIL
+                value: jenkins-x@googlegroups.com
+              - name: GIT_AUTHOR_EMAIL
+                value: jenkins-x@googlegroups.com
+              - name: GIT_AUTHOR_NAME
+                value: jenkins-x-bot
+              - name: GIT_COMMITTER_NAME
+                value: jenkins-x-bot
+              - name: BASE_WORKSPACE
+                value: /workspace/source
+              - name: GOPROXY
+                value: http://jenkins-x-athens-proxy:80
+              - name: PARALLEL_BUILDS
+                value: "2"
+              - name: GHE_ACCESS_TOKEN
+                valueFrom:
+                  secretKeyRef:
+                    name: jx-pipeline-git-github-ghe
+                    key: password
+              - name: JENKINS_PASSWORD
+                valueFrom:
+                  secretKeyRef:
+                    name: test-jenkins-user 
+                    key: password
+            steps:
+              - name: boot-lh-ghe-e2e-tests
+                image: gcr.io/jenkinsxio/builder-go-nodejs:${inputs.params.version}
+                command: ./jx/bdd/boot-lh-ghe/ci.sh
+
+              - name: generate-report
+                image: gcr.io/jenkinsxio/builder-nodejs10x:0.1.607
+                command: jx
+                args:
+                  - step
+                  - report
+                  - --in-dir
+                  - /workspace/source/build/reports
+                  - --target-report
+                  - create_spring_application.junit.xml
+                  - --merge
+                  - --out-dir
+                  - /workspace/source/build/reports
+                  - --output-name
+                  - junit.html
+                  - --suite-name
+                  - "BDD_Boot_LH_GHE_Tests"
+
+              - name: clear-kubeconfig
+                image: gcr.io/jenkinsxio/builder-go:${inputs.params.version}
+                command: rm ~/.kube/config
+
+              - name: stash_html_report
+                image: gcr.io/jenkinsxio/builder-go:${inputs.params.version}
+                command: jx
+                args:
+                  - step
+                  - stash
+                  - -c
+                  - e2e-tests
+                  - -p
+                  - "/workspace/source/build/reports/junit.html"
+                  - --bucket-url gs://jx-prod-logs

--- a/jx/bdd/boot-lh-ghe/README.md
+++ b/jx/bdd/boot-lh-ghe/README.md
@@ -1,0 +1,1 @@
+## BDD test using JX Boot with Local secrets

--- a/jx/bdd/boot-lh-ghe/ci.sh
+++ b/jx/bdd/boot-lh-ghe/ci.sh
@@ -1,0 +1,77 @@
+#!/usr/bin/env bash
+set -e
+set -x
+
+export GHE_USERNAME="dev1"
+export GHE_EMAIL="jenkins-x@googlegroups.com"
+
+# fix broken `BUILD_NUMBER` env var
+export BUILD_NUMBER="$BUILD_ID"
+
+JX_HOME="/tmp/jxhome"
+KUBECONFIG="/tmp/jxhome/config"
+
+# lets avoid the git/credentials causing confusion during the test
+export XDG_CONFIG_HOME=$JX_HOME
+
+mkdir -p $JX_HOME/git
+
+jx --version
+
+# replace the credentials file with a single user entry
+echo "https://$GHE_USERNAME:$GHE_ACCESS_TOKEN@github.beescloud.com" > $JX_HOME/git/credentials
+
+gcloud auth activate-service-account --key-file $GKE_SA
+
+# lets setup git
+git config --global --add user.name JenkinsXBot
+git config --global --add user.email jenkins-x@googlegroups.com
+
+echo "running the BDD tests with JX_HOME = $JX_HOME"
+
+sed -e s/\$VERSION/${VERSION_PREFIX}${VERSION}/g -e s/\$CODECOV_TOKEN/${CODECOV_TOKEN}/g boot-vault.platform.yaml.template > boot-vault.platform.yaml
+
+# setup jx boot parameters
+export JX_VALUE_ADMINUSER_PASSWORD="$JENKINS_PASSWORD"
+export JX_VALUE_PIPELINEUSER_USERNAME="$GHE_USERNAME"
+export JX_VALUE_PIPELINEUSER_EMAIL="$GHE_EMAIL"
+export JX_VALUE_PIPELINEUSER_TOKEN="$GHE_ACCESS_TOKEN"
+export JX_VALUE_PROW_HMACTOKEN="$GHE_ACCESS_TOKEN"
+
+# TODO temporary hack until the batch mode in jx is fixed...
+export JX_BATCH_MODE="true"
+
+# Use the latest boot config promoted in the version stream instead of master to avoid conflicts during boot, because
+# boot fetches always the latest version available in the version stream.
+git clone  https://github.com/jenkins-x/jenkins-x-versions.git versions
+export BOOT_CONFIG_VERSION=$(jx step get dependency-version --host=github.com --owner=jenkins-x --repo=jenkins-x-boot-config --dir versions | sed 's/.*: \(.*\)/\1/')
+git clone https://github.com/jenkins-x/jenkins-x-boot-config.git boot-source
+cd boot-source
+git checkout tags/v${BOOT_CONFIG_VERSION} -b latest-boot-config
+
+cp ../jx/bdd/boot-lh-ghe/jx-requirements.yml .
+cp ../jx/bdd/boot-lh-ghe/parameters.yaml env
+
+cp env/jenkins-x-platform/values.tmpl.yaml tmp.yaml
+cat tmp.yaml ../boot-vault.platform.yaml > env/jenkins-x-platform/values.tmpl.yaml
+rm tmp.yaml
+
+# TODO hack until we fix boot to do this too!
+helm init --client-only
+helm repo add jenkins-x https://storage.googleapis.com/chartmuseum.jenkins-x.io
+
+
+jx step bdd \
+    --versions-repo https://github.com/jenkins-x/jenkins-x-versions.git \
+    --config ../jx/bdd/boot-lh-ghe/cluster.yaml \
+    --gopath /tmp \
+    --git-provider=github \
+    --git-provider-url https://github.beescloud.com \
+    --git-username $GHE_USERNAME \
+    --git-api-token $GHE_ACCESS_TOKEN \
+    --default-admin-password $JENKINS_PASSWORD \
+    --no-delete-app \
+    --no-delete-repo \
+    --tests install \
+    --tests test-quickstart-golang-http \
+    --tests test-app-lifecycle

--- a/jx/bdd/boot-lh-ghe/cluster.yaml
+++ b/jx/bdd/boot-lh-ghe/cluster.yaml
@@ -1,0 +1,18 @@
+clusters:
+  - name: boot-lh-ghe
+    args:
+      - create
+      - cluster
+      - gke
+      - --project-id=jenkins-x-bdd3
+      - -m=n1-standard-2
+      - --min-num-nodes=3
+      - --max-num-nodes=5
+      - -z=europe-west1-c
+      - --skip-login
+      - --skip-installation
+    commands:
+      - command: jx
+        args:
+          - boot
+          - -b

--- a/jx/bdd/boot-lh-ghe/jx-requirements.yml
+++ b/jx/bdd/boot-lh-ghe/jx-requirements.yml
@@ -1,0 +1,46 @@
+gitops: true
+cluster:
+  clusterName: bdd-boot-lh-ghe
+  environmentGitOwner: dev1
+  gitKind: github
+  gitName: ghe
+  gitServer: https://github.beescloud.com
+  project: jenkins-x-bdd3
+  provider: gke
+  zone: europe-west1-c
+environments:
+  - key: dev
+    owner: ""
+    repository: ""
+  - key: staging
+    owner: ""
+    repository: ""
+  - key: production
+    owner: ""
+    repository: ""
+ingress:
+  domain: ""
+  externalDNS: false
+  tls:
+    email: ""
+    enabled: false
+    production: false
+kaniko: true
+secretStorage: vault
+storage:
+  logs:
+    enabled: false
+    url: ""
+  reports:
+    enabled: false
+    url: ""
+  repository:
+    enabled: false
+    url: ""
+webhook: lighthouse
+vault:
+  disableURLDiscovery: true
+versionStream:
+  ref: "master"
+  url: https://github.com/jenkins-x/jenkins-x-versions.git
+

--- a/jx/bdd/boot-lh-ghe/parameters.yaml
+++ b/jx/bdd/boot-lh-ghe/parameters.yaml
@@ -1,0 +1,8 @@
+adminUser:
+  username: admin
+enableDocker: false
+gpg: {}
+pipelineUser:
+  username: dev1
+  email: jenkins-x@googlegroups.com
+

--- a/pkg/cmd/opts/quickstarts.go
+++ b/pkg/cmd/opts/quickstarts.go
@@ -64,7 +64,18 @@ func (o *CommonOptions) LoadQuickStartsFromLocations(locations []v1.QuickStartLo
 			if kind == "" {
 				kind = gits.KindGitHub
 			}
-			gitProvider, err := o.GitProviderForGitServerURL(gitURL, kind, "")
+
+			var gitProvider gits.GitProvider
+			var err error
+
+			// If this is a default quickstart location but there's no github.com credentials, fall back to anonymous
+			server := config.GetOrCreateServer(gitURL)
+			userAuth := config.CurrentUser(server, o.InCluster())
+			if kube.IsDefaultQuickstartLocation(location) && (userAuth == nil || userAuth.IsInvalid()) {
+				gitProvider, err = gits.NewAnonymousGitHubProvider(server, nil)
+			} else {
+				gitProvider, err = o.GitProviderForGitServerURL(gitURL, kind, "")
+			}
 			if err != nil {
 				return model, err
 			}

--- a/pkg/cmd/step/verify/step_verify_environments.go
+++ b/pkg/cmd/step/verify/step_verify_environments.go
@@ -402,7 +402,11 @@ func (o *StepVerifyEnvironmentsOptions) handleDevEnvironmentRepository(envGitInf
 }
 
 func (o *StepVerifyEnvironmentsOptions) createDevEnvironmentRepository(gitInfo *gits.GitRepository, localRepoDir string, fromGitURL string, fromGitRef string, privateRepo bool, requirements *config.RequirementsConfig, provider gits.GitProvider, gitter gits.Gitter) (*gits.GitRepository, error) {
-	if fromGitURL == config.DefaultBootRepository && fromGitRef == "master" {
+	isDefaultBootURL, err := gits.IsDefaultBootConfigURL(fromGitURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to verify whether %s is the default boot config repository", fromGitURL)
+	}
+	if isDefaultBootURL && fromGitRef == "master" {
 		// If the GitURL is not overridden and the GitRef is set to it's default value then look up the version number
 		resolver, err := o.CreateVersionResolver(requirements.VersionStream.URL, requirements.VersionStream.Ref)
 		if err != nil {
@@ -428,7 +432,27 @@ func (o *StepVerifyEnvironmentsOptions) createDevEnvironmentRepository(gitInfo *
 		log.Logger().Debugf("set commitish to '%s'", commitish)
 	}
 
-	duplicateInfo, err := gits.DuplicateGitRepoFromCommitish(gitInfo.Organisation, gitInfo.Name, fromGitURL, commitish, "master", privateRepo, provider, gitter)
+	var fromProvider gits.GitProvider
+	// If the to URL isn't github.com, and the fromGitURL is the default boot repository, use a non-authenticated github.com provider for the from provider.
+	if !gitInfo.IsGitHub() && isDefaultBootURL {
+		fromGitInfo, err := gits.ParseGitURL(fromGitURL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse upstream boot config URL %s", fromGitURL)
+		}
+		if fromGitInfo.IsGitHub() {
+			ghServer := &auth.AuthServer{
+				URL:  "https://github.com",
+				Name: "gh",
+				Kind: "github",
+			}
+			fromProvider, err = gits.NewAnonymousGitHubProvider(ghServer, gitter)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to configure anonymous GitHub provider for upstream boot config URL %s", fromGitURL)
+			}
+		}
+	}
+
+	duplicateInfo, err := gits.DuplicateGitRepoFromCommitish(gitInfo.Organisation, gitInfo.Name, fromGitURL, commitish, "master", privateRepo, provider, gitter, fromProvider)
 	if err != nil {
 		return nil, errors.Wrapf(err, "duplicating %s to %s/%s", fromGitURL, gitInfo.Organisation, gitInfo.Name)
 	}

--- a/pkg/config/install_requirements.go
+++ b/pkg/config/install_requirements.go
@@ -3,7 +3,10 @@ package config
 import (
 	"encoding/json"
 	"fmt"
+	"io/ioutil"
 	"os"
+	"path/filepath"
+	"reflect"
 	"strings"
 
 	"github.com/vrischmann/envconfig"
@@ -16,10 +19,6 @@ import (
 	"github.com/jenkins-x/jx/pkg/cloud"
 	"github.com/jenkins-x/jx/pkg/log"
 	"github.com/pkg/errors"
-
-	"io/ioutil"
-	"path/filepath"
-	"reflect"
 
 	"github.com/jenkins-x/jx/pkg/util"
 )

--- a/pkg/gits/github.go
+++ b/pkg/gits/github.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"io/ioutil"
+	"net/http"
 	"os"
 	"strconv"
 	"strings"
@@ -48,8 +49,26 @@ func NewGitHubProvider(server *auth.AuthServer, user *auth.UserAuth, git Gitter)
 	)
 	tc := oauth2.NewClient(ctx, ts)
 
+	return newGitHubProviderFromOauthClient(tc, provider)
+}
+
+// NewAnonymousGitHubProvider returns a new GitHubProvider without any authentication
+func NewAnonymousGitHubProvider(server *auth.AuthServer, git Gitter) (GitProvider, error) {
+	ctx := context.Background()
+
+	provider := GitHubProvider{
+		Server:  *server,
+		User:    auth.UserAuth{},
+		Context: ctx,
+		Git:     git,
+	}
+
+	return newGitHubProviderFromOauthClient(nil, provider)
+}
+
+func newGitHubProviderFromOauthClient(tc *http.Client, provider GitHubProvider) (GitProvider, error) {
 	var err error
-	u := server.URL
+	u := provider.Server.URL
 	if IsGitHubServerURL(u) {
 		provider.Client = github.NewClient(tc)
 	} else {

--- a/pkg/gits/helpers.go
+++ b/pkg/gits/helpers.go
@@ -8,6 +8,7 @@ import (
 	"os/user"
 	"path"
 	"path/filepath"
+	"reflect"
 	"sort"
 	"strings"
 
@@ -15,9 +16,9 @@ import (
 
 	uuid "github.com/satori/go.uuid"
 
-	"github.com/jenkins-x/jx/pkg/util"
-
+	jxconfig "github.com/jenkins-x/jx/pkg/config"
 	"github.com/jenkins-x/jx/pkg/log"
+	"github.com/jenkins-x/jx/pkg/util"
 
 	"github.com/pkg/errors"
 )
@@ -776,7 +777,7 @@ func FindTagForVersion(dir string, version string, gitter Gitter) (string, error
 // DuplicateGitRepoFromCommitish will duplicate branches (but not tags) from fromGitURL to toOrg/toName. It will reset the
 // head of the toBranch on the duplicated repo to fromCommitish. It returns the GitRepository for the duplicated repo.
 // If the repository already exist and error is returned.
-func DuplicateGitRepoFromCommitish(toOrg string, toName string, fromGitURL string, fromCommitish string, toBranch string, private bool, provider GitProvider, gitter Gitter) (*GitRepository, error) {
+func DuplicateGitRepoFromCommitish(toOrg string, toName string, fromGitURL string, fromCommitish string, toBranch string, private bool, provider GitProvider, gitter Gitter, fromProvider GitProvider) (*GitRepository, error) {
 	log.Logger().Debugf("getting repo %s/%s", toOrg, toName)
 	_, err := provider.GetRepository(toOrg, toName)
 	if err == nil {
@@ -789,10 +790,16 @@ func DuplicateGitRepoFromCommitish(toOrg string, toName string, fromGitURL strin
 	if err != nil {
 		return nil, errors.Wrapf(err, "parsing %s", fromGitURL)
 	}
-	fromInfo, err = provider.GetRepository(fromInfo.Organisation, fromInfo.Name)
+
+	// If no separate fromProvider is specified, use the same provider.
+	if fromProvider == nil || reflect.ValueOf(fromProvider).IsNil() {
+		fromProvider = provider
+	}
+	fromInfo, err = fromProvider.GetRepository(fromInfo.Organisation, fromInfo.Name)
 	if err != nil {
 		return nil, errors.Wrapf(err, "getting repo for %s", fromGitURL)
 	}
+
 	duplicateInfo, err := provider.CreateRepository(toOrg, toName, private)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to create GitHub repo %s/%s", toOrg, toName)
@@ -944,4 +951,17 @@ func RefIsBranch(dir string, ref string, gitter Gitter) (bool, error) {
 		}
 	}
 	return false, nil
+}
+
+// IsDefaultBootConfigURL checks if the given URL corresponds to the default boot config URL
+func IsDefaultBootConfigURL(url string) (bool, error) {
+	gitInfo, err := ParseGitURL(url)
+	if err != nil {
+		return false, errors.Wrap(err, "couldn't parse provided repo URL")
+	}
+	defaultInfo, err := ParseGitURL(jxconfig.DefaultBootRepository)
+	if err != nil {
+		return false, errors.Wrap(err, "couldn't parse default boot config URL")
+	}
+	return gitInfo.HttpsURL() == defaultInfo.HttpsURL(), nil
 }

--- a/pkg/gits/helpers_test.go
+++ b/pkg/gits/helpers_test.go
@@ -9,6 +9,7 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/jenkins-x/jx/pkg/config"
 	"github.com/jenkins-x/jx/pkg/util"
 
 	"github.com/jenkins-x/jx/pkg/tests"
@@ -1376,6 +1377,14 @@ func TestDuplicateGitRepoFromCommitish(t *testing.T) {
 		return nil
 	}, gitter)
 	assert.NoError(t, err)
+	otherProviderRepo, err := gits.NewFakeRepository("foo", "bar", func(dir string) error {
+		err := ioutil.WriteFile(filepath.Join(dir, "README"), []byte("Goodbye!"), 0655)
+		if err != nil {
+			return errors.Wrapf(err, "writing README")
+		}
+		return nil
+	}, gitter)
+	assert.NoError(t, err)
 
 	dir, err := ioutil.TempDir("", "")
 	assert.NoError(t, err)
@@ -1432,12 +1441,13 @@ func TestDuplicateGitRepoFromCommitish(t *testing.T) {
 		gitter        gits.Gitter
 	}
 	tests := []struct {
-		provider  *gits.FakeProvider
-		name      string
-		args      args
-		want      *gits.GitRepository
-		wantFiles map[string][]byte
-		wantErr   string
+		provider         *gits.FakeProvider
+		useOtherProvider bool
+		name             string
+		args             args
+		want             *gits.GitRepository
+		wantFiles        map[string][]byte
+		wantErr          string
 	}{
 		{
 			name: "sameOrg",
@@ -1497,6 +1507,37 @@ func TestDuplicateGitRepoFromCommitish(t *testing.T) {
 			},
 			wantFiles: map[string][]byte{
 				"README": []byte("Hello!"),
+			},
+		},
+		{
+			name: "differentProvider",
+			args: args{
+				toOrg:         "coyote",
+				toName:        "wile",
+				fromGitURL:    "https://fake.git/foo/bar.git",
+				fromCommitish: "master",
+				toBranch:      "master",
+				gitter:        gitter,
+			},
+			useOtherProvider: true,
+			want: &gits.GitRepository{
+				Name:             "wile",
+				AllowMergeCommit: false,
+				HTMLURL:          "https://fake.git/coyote/wile",
+				CloneURL:         "",
+				SSHURL:           "",
+				Language:         "",
+				Fork:             false,
+				Stars:            0,
+				URL:              "https://fake.git/coyote/wile.git",
+				Scheme:           "https",
+				Host:             "fake.git",
+				Organisation:     "coyote",
+				Project:          "",
+				Private:          false,
+			},
+			wantFiles: map[string][]byte{
+				"README": []byte("Goodbye!"),
 			},
 		},
 		{
@@ -1621,6 +1662,38 @@ func TestDuplicateGitRepoFromCommitish(t *testing.T) {
 				"LICENSE": []byte("TODO"),
 			},
 		},
+		{
+			name: "wrongDifferentProvider",
+			args: args{
+				toOrg:         "coyote",
+				toName:        "wile",
+				fromGitURL:    "https://fake.git/acme/roadrunner.git",
+				fromCommitish: "master",
+				toBranch:      "master",
+				gitter:        gitter,
+			},
+			useOtherProvider: true,
+			want: &gits.GitRepository{
+				Name:             "wile",
+				AllowMergeCommit: false,
+				HTMLURL:          "https://fake.git/coyote/wile",
+				CloneURL:         "",
+				SSHURL:           "",
+				Language:         "",
+				Fork:             false,
+				Stars:            0,
+				URL:              "https://fake.git/coyote/wile.git",
+				Scheme:           "https",
+				Host:             "fake.git",
+				Organisation:     "coyote",
+				Project:          "",
+				Private:          false,
+			},
+			wantErr: "organization 'acme' not found",
+			wantFiles: map[string][]byte{
+				"README": []byte("Hello!"),
+			},
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
@@ -1635,7 +1708,13 @@ func TestDuplicateGitRepoFromCommitish(t *testing.T) {
 			}
 			tt.provider = provider
 
-			got, err := gits.DuplicateGitRepoFromCommitish(tt.args.toOrg, tt.args.toName, tt.args.fromGitURL, tt.args.fromCommitish, tt.args.toBranch, false, tt.provider, tt.args.gitter)
+			var fromProvider *gits.FakeProvider
+			if tt.useOtherProvider {
+				fromProvider = gits.NewFakeProvider(otherProviderRepo)
+				fromProvider.Gitter = gitter
+			}
+
+			got, err := gits.DuplicateGitRepoFromCommitish(tt.args.toOrg, tt.args.toName, tt.args.fromGitURL, tt.args.fromCommitish, tt.args.toBranch, false, tt.provider, tt.args.gitter, fromProvider)
 			if tt.wantErr != "" {
 				assert.Error(t, err)
 				assert.Contains(t, err.Error(), tt.wantErr)
@@ -1701,7 +1780,7 @@ func Test_DuplicateGitRepoFromCommitish_returns_error_if_target_repo_exists(t *t
 	provider := gits.NewFakeProvider(originalRepo, targetRepo)
 	provider.Gitter = gitter
 
-	repo, err := gits.DuplicateGitRepoFromCommitish(targetRepo.GitRepo.Organisation, targetRepo.GitRepo.Name, originalRepo.GitRepo.CloneURL, "origin/foo", "bar", false, provider, gitter)
+	repo, err := gits.DuplicateGitRepoFromCommitish(targetRepo.GitRepo.Organisation, targetRepo.GitRepo.Name, originalRepo.GitRepo.CloneURL, "origin/foo", "bar", false, provider, gitter, nil)
 	assert.Error(t, err)
 	assert.Equal(t, "repository acme/coyote already exists", err.Error())
 	assert.Nil(t, repo)
@@ -2394,4 +2473,31 @@ func TestIsCouldntFindRemoteRefErrorHandlesUppercaseRef(t *testing.T) {
 	error := errors.New(" fatal: couldn't find remote ref add-app-your-app-0.0.0-SNAPSHOT-PR-1234-1:")
 	ref := "add-app-your-app-0.0.0-SNAPSHOT-PR-1234-1"
 	assert.True(t, gits.IsCouldntFindRemoteRefError(error, ref))
+}
+
+func TestIsDefaultBootConfigURL(t *testing.T) {
+	wrongURL := "https://github.com/something-else/jenkins-x-boot-config.git"
+
+	var rightURLWithDotGit string
+	var rightURLWithoutDotGit string
+
+	if strings.HasSuffix(config.DefaultBootRepository, ".git") {
+		rightURLWithDotGit = config.DefaultBootRepository
+		rightURLWithoutDotGit = strings.TrimSuffix(config.DefaultBootRepository, ".git")
+	} else {
+		rightURLWithoutDotGit = config.DefaultBootRepository
+		rightURLWithDotGit = config.DefaultBootRepository + ".git"
+	}
+
+	withWrongURL, err := gits.IsDefaultBootConfigURL(wrongURL)
+	assert.NoError(t, err)
+	assert.False(t, withWrongURL)
+
+	withRightURLDotGit, err := gits.IsDefaultBootConfigURL(rightURLWithDotGit)
+	assert.NoError(t, err)
+	assert.True(t, withRightURLDotGit)
+
+	withRightURLNoDotGit, err := gits.IsDefaultBootConfigURL(rightURLWithoutDotGit)
+	assert.NoError(t, err)
+	assert.True(t, withRightURLNoDotGit)
 }

--- a/pkg/kube/quickstarts.go
+++ b/pkg/kube/quickstarts.go
@@ -2,6 +2,7 @@ package kube
 
 import (
 	"fmt"
+	"reflect"
 
 	v1 "github.com/jenkins-x/jx/pkg/apis/jenkins.io/v1"
 	"github.com/jenkins-x/jx/pkg/client/clientset/versioned"
@@ -33,4 +34,14 @@ func GetQuickstartLocations(jxClient versioned.Interface, ns string) ([]v1.Quick
 	}
 	answer = env.Spec.TeamSettings.QuickstartLocations
 	return answer, nil
+}
+
+// IsDefaultQuickstartLocation checks whether the given quickstart location is a default one, and if so returns true
+func IsDefaultQuickstartLocation(location v1.QuickStartLocation) bool {
+	for _, l := range DefaultQuickstartLocations {
+		if reflect.DeepEqual(l, location) {
+			return true
+		}
+	}
+	return false
 }

